File: queryable-encryption/dotnet/gcp/reader/QueryableEncryption/MakeDataKey.cs
﻿using System;
using System.IO;
using System.Collections.Generic;
using System.Threading;
using MongoDB.Driver;
using MongoDB.Bson;
using MongoDB.Driver.Encryption;
using Credentials;

namespace QueryableEncryption
{
    internal static class MakeDataKey
    {
        public static void MakeKey()
        {
            var credentials = new YourCredentials().GetCredentials();

            // start-kmsproviders
            var kmsProviders = new Dictionary<string, IReadOnlyDictionary<string, object>>();
            const string provider = "gcp";
            var gcpPrivateKey = credentials["GCP_PRIVATE_KEY"];
            var gcpEmail = credentials["GCP_EMAIL"];
            var gcpKmsOptions = new Dictionary<string, object>
            {
               { "privateKey", gcpPrivateKey },
               { "email", gcpEmail },
            };
            kmsProviders.Add(provider, gcpKmsOptions);

            var gcpDataKeyProjectId = credentials["GCP_PROJECT_ID"];
            var gcpDataKeyLocation = credentials["GCP_LOCATION"];
            var gcpDataKeyKeyRing = credentials["GCP_KEY_RING"];
            var gcpDataKeyKeyName = credentials["GCP_KEY_NAME"];
            // end-kmsproviders

            // start-datakeyopts
            DataKeyOptions GetDataKeyOptions(List<string> altNames)
            {
                var dataKeyOptions = new DataKeyOptions(
                   alternateKeyNames: altNames,
                   masterKey: new BsonDocument
                   {
                       { "projectId", gcpDataKeyProjectId },
                       { "location", gcpDataKeyLocation } ,
                       { "keyRing", gcpDataKeyKeyRing },
                       { "keyName", gcpDataKeyKeyName },
                   });
                return dataKeyOptions;
            }
            // end-datakeyopts

            // start-create-index
            var connectionString = credentials["MONGODB_URI"];
            var keyVaultNamespace = CollectionNamespace.FromFullName("encryption.__keyVault");
            var keyVaultClient = new MongoClient(connectionString);
            var indexOptions = new CreateIndexOptions<BsonDocument>
            {
                Unique = true,
                PartialFilterExpression = new BsonDocument
                    {{"keyAltNames", new BsonDocument {{"$exists", new BsonBoolean(true)}}}}
            };
            var builder = Builders<BsonDocument>.IndexKeys;
            var indexKeysDocument = builder.Ascending("keyAltNames");
            var indexModel = new CreateIndexModel<BsonDocument>(indexKeysDocument, indexOptions);
            var keyVaultDatabase = keyVaultClient.GetDatabase(keyVaultNamespace.DatabaseNamespace.DatabaseName);
            // Drop the Key Vault Collection in case you created this collection
            // in a previous run of this application.
            keyVaultDatabase.DropCollection(keyVaultNamespace.CollectionName);
            var keyVaultCollection = keyVaultDatabase.GetCollection<BsonDocument>(keyVaultNamespace.CollectionName);
            keyVaultCollection.Indexes.CreateOne(indexModel);
            // end-create-index

            // start-create-dek
            var clientEncryptionOptions = new ClientEncryptionOptions(
                keyVaultClient,
                keyVaultNamespace,
                kmsProviders);
            var clientEncryption = new ClientEncryption(clientEncryptionOptions);
            var dataKeyOptions1 = GetDataKeyOptions(new List<string> { "dataKey1" });
            var dataKeyOptions2 = GetDataKeyOptions(new List<string> { "dataKey2" });
            var dataKeyOptions3 = GetDataKeyOptions(new List<string> { "dataKey3" });
            var dataKeyOptions4 = GetDataKeyOptions(new List<string> { "dataKey4" });


            BsonBinaryData CreateKeyGetID(DataKeyOptions options)
            {
                var dateKeyGuid = clientEncryption.CreateDataKey(provider, options, CancellationToken.None);
                return new BsonBinaryData(dateKeyGuid, GuidRepresentation.Standard);
            }

            var dataKeyId1 = CreateKeyGetID(dataKeyOptions1);
            var dataKeyId2 = CreateKeyGetID(dataKeyOptions2);
            var dataKeyId3 = CreateKeyGetID(dataKeyOptions3);
            var dataKeyId4 = CreateKeyGetID(dataKeyOptions4);
            // end-create-dek

            // start-create-enc-collection
            var encryptedCollectionNamespace = CollectionNamespace.FromFullName("medicalRecords.patients");
            var encryptedFieldsMap = new Dictionary<string, BsonDocument>
            {
                {
                    encryptedCollectionNamespace.FullName, new BsonDocument
                    {
                        {
                            "fields", new BsonArray
                            {
                                new BsonDocument
                                {
                                    {"keyId", dataKeyId1},
                                    {"path", new BsonString("patientId")},
                                    {"bsonType", new BsonString("int")},
                                    {
                                        "queries", new BsonDocument
                                        {
                                            {"queryType", new BsonString("equality")}
                                        }
                                    }
                                },
                                new BsonDocument
                                {
                                    {"keyId", dataKeyId2},
                                    {"path", new BsonString("medications")},
                                    {"bsonType", new BsonString("array")},
                                },
                                new BsonDocument
                                {
                                    {"keyId", dataKeyId3},
                                    {"path", new BsonString("patientRecord.ssn")},
                                    {"bsonType", new BsonString("string")},
                                    {
                                        "queries", new BsonDocument
                                        {
                                            {"queryType", new BsonString("equality")}
                                        }
                                    }
                                },
                                new BsonDocument
                                {
                                    {"keyId", dataKeyId4},
                                    {"path", new BsonString("patientRecord.billing")},
                                    {"bsonType", new BsonString("object")},
                                },
                            }
                        }
                    }
                }
            };

            var extraOptions = new Dictionary<string, object>()
            {
                {"cryptSharedLibPath", credentials["SHARED_LIB_PATH"]},
            };

            var autoEncryptionOptions = new AutoEncryptionOptions(
                keyVaultNamespace,
                kmsProviders,
                encryptedFieldsMap: encryptedFieldsMap,
                extraOptions: extraOptions);

            var clientSettings = MongoClientSettings.FromConnectionString(connectionString);
            clientSettings.AutoEncryptionOptions = autoEncryptionOptions;
            var secureClient = new MongoClient(clientSettings);
            var encryptedDatabase = secureClient.GetDatabase(encryptedCollectionNamespace.DatabaseNamespace.DatabaseName);
            // Drop the encrypted collection in case you created this collection
            // in a previous run of this application.
            encryptedDatabase.DropCollection(encryptedCollectionNamespace.CollectionName);
            encryptedDatabase.CreateCollection(encryptedCollectionNamespace.CollectionName);
            Console.WriteLine("Created encrypted collection!");
            // end-create-enc-collection
        }
    }
}

File: queryable-encryption/node/azure/reader/make_data_key.js
const { MongoClient, Binary } = require("mongodb");
const { ClientEncryption } = require("mongodb-client-encryption");

const keyVaultDatabase = "encryption";
const keyVaultCollection = "__keyVault";
const keyVaultNamespace = `${keyVaultDatabase}.${keyVaultCollection}`;
const secretDB = "medicalRecords";
const secretCollection = "patients";

const { getCredentials } = require("./your_credentials");
credentials = getCredentials();

// start-kmsproviders
const provider = "azure";
const kmsProviders = {
  azure: {
    tenantId: credentials["AZURE_TENANT_ID"],
    clientId: credentials["AZURE_CLIENT_ID"],
    clientSecret: credentials["AZURE_CLIENT_SECRET"],
  },
};
// end-kmsproviders

// start-datakeyopts
const masterKey = {
  keyVaultEndpoint: credentials["AZURE_KEY_VAULT_ENDPOINT"],
  keyName: credentials["AZURE_KEY_NAME"],
};
// end-datakeyopts

async function run() {
  // start-create-index
  const uri = credentials.MONGODB_URI;
  const keyVaultClient = new MongoClient(uri);
  await keyVaultClient.connect();
  const keyVaultDB = keyVaultClient.db(keyVaultDatabase);
  // Drop the Key Vault Collection in case you created this collection
  // in a previous run of this application.
  await keyVaultDB.dropDatabase();
  const keyVaultColl = keyVaultDB.collection(keyVaultCollection);
  await keyVaultColl.createIndex(
    { keyAltNames: 1 },
    {
      unique: true,
      partialFilterExpression: { keyAltNames: { $exists: true } },
    }
  );
  // end-create-index

  // start-create-dek
  const clientEnc = new ClientEncryption(keyVaultClient, {
    keyVaultNamespace: keyVaultNamespace,
    kmsProviders: kmsProviders,
  });
  const dek1 = await clientEnc.createDataKey(provider, {
    masterKey: masterKey,
    keyAltNames: ["dataKey1"],
  });
  const dek2 = await clientEnc.createDataKey(provider, {
    masterKey: masterKey,
    keyAltNames: ["dataKey2"],
  });
  const dek3 = await clientEnc.createDataKey(provider, {
    masterKey: masterKey,
    keyAltNames: ["dataKey3"],
  });
  const dek4 = await clientEnc.createDataKey(provider, {
    masterKey: masterKey,
    keyAltNames: ["dataKey4"],
  });
  // end-create-dek

  // start-create-enc-collection
  const encryptedFieldsMap = {
    [`${secretDB}.${secretCollection}`]: {
      fields: [
        {
          keyId: dek1,
          path: "patientId",
          bsonType: "int",
          queries: { queryType: "equality" },
        },
        {
          keyId: dek2,
          path: "medications",
          bsonType: "array",
        },
        {
          keyId: dek3,
          path: "patientRecord.ssn",
          bsonType: "string",
          queries: { queryType: "equality" },
        },
        {
          keyId: dek4,
          path: "patientRecord.billing",
          bsonType: "object",
        },
      ],
    },
  };
  const extraOptions = {
    cryptSharedLibPath: credentials["SHARED_LIB_PATH"],
  };
  const encClient = new MongoClient(uri, {
    autoEncryption: {
      keyVaultNamespace,
      kmsProviders,
      extraOptions,
      encryptedFieldsMap,
    },
  });
  await encClient.connect();
  const newEncDB = encClient.db(secretDB);
  // Drop the encrypted collection in case you created this collection
  // in a previous run of this application.
  await newEncDB.dropDatabase();
  await newEncDB.createCollection(secretCollection);
  console.log("Created encrypted collection!");
  // end-create-enc-collection
  await keyVaultClient.close();
  await encClient.close();
}

run().catch(console.dir);
